pkg/controllers/terminating/v1alpha1/terminate.go

```diff
@@ -62,7 +62,8 @@ func (t *Terminator) drain(ctx context.Context, node *v1.Node) (bool, error) {
 	// 2. Evict pods on node
 	empty := true
 	for _, p := range pods {
-		if !pod.IsOwnedByDaemonSet(p) {
+		// Don't evict pods that will re-schedule after being evicted
+		if !pod.ToleratesUnschedulable(&p.Spec) {
```

Comment:
I'm wondering how this works? https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-nodes-by-condition suggests (I'm probably not understanding) that daemons usually do tolerate unschedulable - so will this really achieve the result of graceful shutdown of daemonsets?

Reply:
The challenge here is that when you evict these pods, the kcm will recreate them, and the kube scheduler will put them right back on the node (since they tolerate the cordon). There's more or less no safe way to drain these pods gracefully without reschedule loops. Luckily https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/#:~:text=With%20graceful%20node%20shutdown%2C%20the,any%20resources%20they%20are%20holding. should handle the cleanup logic as part of instance termination (assuming AWS EC2 isn't too aggressive).

Reply:
With regards to daemonsets, all daemonsets will have the unschedulable taint [link](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations).

My understanding is that daemonsets are commonly processes that are essential to the operation of the node, so when the instance/node terminates, the daemonsets will also be shut down.

In addition, I changed the code to drain non-critical pods first, then critical pods to go with https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown